src/tools/util.c

```diff
@@ -139,24 +137,36 @@ util_connect_card_ex(sc_context_t *ctx, sc_card_t **cardp,
 				errno = 0;
 				num = strtol(reader_id, &endptr, 0);
 				if (!errno && endptr && *endptr == '\0')
-					reader = sc_ctx_get_reader(ctx, num);
+					*reader = sc_ctx_get_reader(ctx, num);
 				else
-					reader = sc_ctx_get_reader_by_name(ctx, reader_id);
+					*reader = sc_ctx_get_reader_by_name(ctx, reader_id);
 			}
 		}
 autofound:
-		if (!reader) {
+		if (!(*reader)) {
 			fprintf(stderr, "Reader \"%s\" not found (%d reader(s) detected)\n",
 					reader_id, sc_ctx_get_reader_count(ctx));
 			return 1;
 		}
 
-		if (sc_detect_card_presence(reader) <= 0) {
+		if (sc_detect_card_presence(*reader) <= 0) {
 			fprintf(stderr, "Card not present.\n");
 			return 3;
 		}
 	}
+	return 0;
+}
+int
+util_connect_card_ex(sc_context_t *ctx, sc_card_t **cardp,
```

Comment:
This function is internal, why not change the signature and not make a new _ex ?

Reply:
The `util_connect_card_ex()`  is function with 6 arguments.  One specific argument is  `int do_lock`.  The `util_connect_card()` function is same as `util_connect_card_ex()` with `do_lock` set to 1.

There is 8 calls in whole opensc sources  to `util_connect_card_ex()` with `do_lock` set to  0,  and 9 calls to `util_connect_card()` (in 18 files).

IMHO, this PR should not change function `util_connect_card_ex ()` (because this introduces changes into files not affected by #1606).  If the `util_connect_card()` and `util_connect_card_ex ()` interface is not satisfactory, it should be solved in separate PR.